pkg/controllers/capr/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,366 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// dynamicGetter defines the interface for the Get method from dynamic.Controller
+type dynamicGetter interface {
+	Get(gvk schema.GroupVersionKind, namespace string, name string) (runtime.Object, error)
+}
+
+type autoscalerHandler struct {
+	capiClusterCache           v1beta1.ClusterCache
+	capiMachineCache           v1beta1.MachineCache
+	capiMachineDeploymentCache v1beta1.MachineDeploymentCache
+
+	clusterClient v2provcontrollers.ClusterClient
+	clusterCache  v2provcontrollers.ClusterCache
+
+	globalRoleClient mgmtcontrollers.GlobalRoleClient
+	globalRoleCache  mgmtcontrollers.GlobalRoleCache
+
+	globalRoleBindingClient mgmtcontrollers.GlobalRoleBindingClient
+	globalRoleBindingCache  mgmtcontrollers.GlobalRoleBindingCache
+
+	userClient mgmtcontrollers.UserClient
+	userCache  mgmtcontrollers.UserCache
+
+	tokenClient mgmtcontrollers.TokenClient
+	tokenCache  mgmtcontrollers.TokenCache
+
+	secretClient wranglerv1.SecretClient
+	secretCache  wranglerv1.SecretCache
+
+	helmOp      fleetcontrollers.HelmOpController
+	helmOpCache fleetcontrollers.HelmOpCache
+
+	dynamicClient dynamicGetter
+}
+
+func Register(ctx context.Context, clients *wrangler.CAPIContext) {
+	h := &autoscalerHandler{
+		clusterClient: clients.Provisioning.Cluster(),
+		clusterCache:  clients.Provisioning.Cluster().Cache(),
+
+		capiClusterCache:           clients.CAPI.Cluster().Cache(),
+		capiMachineCache:           clients.CAPI.Machine().Cache(),
+		capiMachineDeploymentCache: clients.CAPI.MachineDeployment().Cache(),
+
+		globalRoleClient:        clients.Mgmt.GlobalRole(),
+		globalRoleCache:         clients.Mgmt.GlobalRole().Cache(),
+		globalRoleBindingClient: clients.Mgmt.GlobalRoleBinding(),
+		globalRoleBindingCache:  clients.Mgmt.GlobalRoleBinding().Cache(),
+
+		userClient: clients.Mgmt.User(),
+		userCache:  clients.Mgmt.User().Cache(),
+
+		tokenClient: clients.Mgmt.Token(),
+		tokenCache:  clients.Mgmt.Token().Cache(),
+
+		secretClient: clients.Core.Secret(),
+		secretCache:  clients.Core.Secret().Cache(),
+
+		helmOp:      clients.Fleet.HelmOp(),
+		helmOpCache: clients.Fleet.HelmOp().Cache(),
+
+		dynamicClient: clients.Dynamic,
+	}
+
+	// only run the "create" handlers if autoscaling is enabled. otherwise only run the cleanup handler
+	// (in case the user disabled autoscaling after having it enabled
+	if !features.ClusterAutoscaling.Enabled() {
+		clients.CAPI.Cluster().OnChange(ctx, "autoscaler-cleanup", h.ensureCleanup)
+		return
+	}
+
+	// warn the user if they have the autoscaling feature-flag enabled but no chart repo set,
+	// then do not run the controller.
+	if settings.ClusterAutoscalerChartRepository.Get() == "" {
+		logrus.Warnf("[autoscaler] no value is set for the cluster-autoscaler-chart-repo Setting  - cannot enable autoscaling!")
```

Comment:
Nothing needs to be changed, just wanted to point out that this warning message appears only during Rancher starts up.